File: android/src/main/java/com/ammarahmed/rnadmob/nativeads/CacheManager.java
package com.ammarahmed.rnadmob.nativeads;

import android.content.Context;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.google.android.gms.ads.AdListener;

import java.util.HashMap;
import java.util.Map;

public class CacheManager {

    Map<String, RNAdMobUnifiedAdQueueWrapper> repositoriesMap = new HashMap<>();

    public static CacheManager instance = new CacheManager();
    public static final String EVENT_AD_PRELOAD_LOADED = "onAdPreloadLoaded";
    public static final String EVENT_AD_PRELOAD_ERROR = "onAdPreloadError";

    public boolean isLoading(String id) {
        if (repositoriesMap.get(id) != null) {
            return repositoriesMap.get(id).isLoading();
        } else {
            return false;
        }
    }

    public int numberOfAds(String id) {
        if (repositoriesMap.containsKey(id)) {
            return repositoriesMap.get(id).nativeAds.size();
        } else {
            return 0;
        }

    }

    public void attachAdListener(String id, AdListener listener) {
        if (repositoriesMap.get(id) != null){
            repositoriesMap.get(id).attachAdListener(listener);
        }
    }

    public void detachAdListener(String id) {
        if (repositoriesMap.get(id) != null){
            repositoriesMap.get(id).detachAdListener();
        }
    }

    public WritableMap registerRepo(Context context, ReadableMap config) {
        try {
            String repo = null;
            WritableMap args = Arguments.createMap();
            if (!config.hasKey("adUnitId") || config.getString("adUnitId") == null){
                args.putBoolean("success", false);
                args.putString("error", "the adUnitId has to be set in config");
            }
            if (config.hasKey("name") && config.getString("name") != null) {
                repo = config.getString("name");
            } else {
                if (config.hasKey("adUnitId") && config.getString("adUnitId") != null) {
                    repo = config.getString("adUnitId");
                }
            }
            if (repo != null){
                if (!repositoriesMap.containsKey(repo)){
                    repositoriesMap.put(repo, new RNAdMobUnifiedAdQueueWrapper(context, config, repo));
                    args.putBoolean("success", true);
                    args.putString("repo", repo);
                } else {
                    args.putBoolean("success", false);
                    args.putString("error", "the given repo has been registered before");
                }
            } else {
                args.putBoolean("success", false);
                args.putString("error", "the adUnitId or name has to be set in config");
            }
            return args;
        } catch (Exception e) {
            WritableMap args = Arguments.createMap();
            args.putBoolean("success", false);
            args.putString("error", e.getCause() != null ? e.getCause().toString() : "");
            return args;
        }
    }

    public void unRegisterRepo(String repo) {
        repositoriesMap.remove(repo);
    }

    public void resetCache(){
        repositoriesMap.clear();
    }

    public void requestAds(String repo){
        repositoriesMap.get(repo).loadAds();
    }

    public void requestAd(String repo){
        repositoriesMap.get(repo).loadAd();
    }

    public Boolean isRegistered(String repo){
        return repositoriesMap.containsKey(repo);
    }

    public RNAdMobUnifiedAdContainer getNativeAd(String repo) {

        if (repositoriesMap.containsKey(repo)) {
            return repositoriesMap.get(repo).getAd();
        } else {
            return null;
        }
    }

    public WritableMap hasAd(String repo) {
        if (repositoriesMap.containsKey(repo)) {
            return repositoriesMap.get(repo).hasAd();
        } else {
            WritableMap args = Arguments.createMap();
            args.putInt(repo, 0);
            return args;
        }
    }

}

File: ios/RNAdMobManager/CacheManager.m
//
//  CacheManager.m
//  react-native-admob-native-ads
//
//  Created by Ali on 8/25/21.
//

#import <Foundation/Foundation.h>
#import <CacheManager.h>
#import "RNAdMobUnifiedAdQueueWrapper.h"



@implementation CacheManager{
    NSMutableDictionary *repositoriesMap;
}

+ (NSString *)EVENT_AD_PRELOAD_LOADED { return @"onAdPreloadLoaded"; }
+ (NSString *)EVENT_AD_PRELOAD_ERROR { return @"onAdPreloadError"; }

static CacheManager *_sharedInstance = nil;

+ (CacheManager*)sharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [[self alloc] init];

    });

    return _sharedInstance;
}
- (instancetype)init{
    repositoriesMap = [[NSMutableDictionary alloc] init];
    return self;
}

-(BOOL) isLoading:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    return  repo && repo.isLoading;
}
-(NSInteger)  numberOfAds:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    if (repo != nil){
      return repo.nativeAds.count;
    }else{
        return 0;
    }
}
-(void) attachAdListener:(NSString*) id listener:(id<AdListener>)listener{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    if (repo != nil){
        [repo attachAdListener:listener];
    }
  
}
-(void) detachAdListener:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    if (repo != nil){
        [repo detachAdListener];
    }
}
-(NSDictionary*)registerRepo:(NSDictionary*) config rootVC:(UIViewController*)rootVC{
    
    NSMutableDictionary*  args = [[NSMutableDictionary alloc] init];
    NSString* repoId = nil;
    if (![config objectForKey:@"adUnitId"]) {
        [args setObject:[NSNumber numberWithBool:false] forKey:@"success"];
        [args setObject:@"the adUnitId has to be set in config" forKey:@"error"];
    }
    
    if ([config objectForKey:@"name"]) {
         repoId = [config objectForKey:@"name"];
    }else{
        if ([config objectForKey:@"adUnitId"]) {
            repoId = [config objectForKey:@"adUnitId"];
        }
    }
    if (repoId != nil){
             
                if (![repositoriesMap objectForKey:repoId]) {
                    RNAdMobUnifiedAdQueueWrapper *repo = [[RNAdMobUnifiedAdQueueWrapper alloc] initWithConfig:config repo:repoId rootVC:rootVC];
                    [repositoriesMap setObject:repo forKey:repoId];
                
                    [args setObject:[NSNumber numberWithBool:YES] forKey:@"success"];
                    [args setObject:repoId forKey:@"repo"];
                }else{
                    [args setObject:[NSNumber numberWithBool:NO] forKey:@"success"];
                    [args setObject:@"the given repo has been registered before" forKey:@"error"];
                }
                
    }else{
        [args setObject:[NSNumber numberWithBool:NO] forKey:@"success"];
        [args setObject:@"the adUnitId or name has to be set in config" forKey:@"error"];
    }
    return  args;
    
}
-(void) unRegisterRepo:(NSString*) id{
    [repositoriesMap removeObjectForKey: id];
}
-(void) resetCache{
    [repositoriesMap removeAllObjects];
}
-(void) configAdLoader:(NSString*) id rootVC:(UIViewController *) rootVC{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    [repo configAdLoaderOption:rootVC];
}
-(void) requestAds:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    [repo fillAds];
}

-(BOOL) isRegistered:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
  return repo;
}
-(RNAdMobUnifiedAdContainer*) getNativeAd:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    if (repo != nil) {
        return [repo getAd];
    }else{
        return nil;
    }
}
-(NSDictionary*) hasAd:(NSString*) id{
    RNAdMobUnifiedAdQueueWrapper *repo =  (RNAdMobUnifiedAdQueueWrapper *)([repositoriesMap objectForKey:id]);
    if (repo != nil) {
        return [repo hasAd];
    }else{
        NSMutableDictionary*  args = [[NSMutableDictionary alloc] init];
        [args setValue:0 forKey:id];
        return  args;
    }
}
@end
